### PR TITLE
Fix form pipe parser

### PR DESCRIPTION
**Pull request by Tratcher**
#### Description
#12381 
Issue 1) Keys without values were being merged "key1&key2=value2"
Issue 2) Keys without values at the end of a form being dropped (2.2) or throwing (3.0) "key1=value1&key2".
		
#### Customer Impact
Externally posted forms were being mis-parsed in some situations.
		
#### Regression?
Yes, in 2.2 un-parsed trailing data would be discarded, but in 3.0 it throws.
		
#### Risk
Low, we have good unit test coverage here.

Old algorithm: Look for "=", then "&".
New algorithm: Look for "&", then see if it has an optional "="

### CODE REVIEW

**Comment by davidfowl** [comment]
Perf numbers?

File: src/Http/WebUtilities/src/FormPipeReader.cs
Comment (davidfowl) [inline src/Http/WebUtilities/src/FormPipeReader.cs]: Very nice change, I was looking at this earlier (if we should swap to the fast path). You can use FirstSpan instead of First.Span (it’s new in 3.0)

File: src/Http/WebUtilities/src/FormPipeReader.cs
Comment (davidfowl) [inline src/Http/WebUtilities/src/FormPipeReader.cs]: Do you need to assign this?

File: src/Http/WebUtilities/src/FormPipeReader.cs
Comment (davidfowl) [inline src/Http/WebUtilities/src/FormPipeReader.cs]: This looks expensive, one sequence reader per loop?

File: src/Http/WebUtilities/src/FormPipeReader.cs
Comment (jkotalik) [inline src/Http/WebUtilities/src/FormPipeReader.cs]: nit: can this be default?

File: src/Http/WebUtilities/src/FormPipeReader.cs
Comment (Tratcher) [reply src/Http/WebUtilities/src/FormPipeReader.cs]: I can't rewind the main reader can I? 

It's a ref struct, how expensive are those?

File: src/Http/WebUtilities/src/FormPipeReader.cs
Comment (jkotalik) [inline src/Http/WebUtilities/src/FormPipeReader.cs]: Super nit: maybe we can share the equals and and delimiters via ref to ParseFormValuesFast/Slow?

File: src/Http/WebUtilities/src/FormPipeReader.cs
Comment (Tratcher) [reply src/Http/WebUtilities/src/FormPipeReader.cs]: No. It's here for clarity and consistency with the fast path.

File: src/Http/WebUtilities/src/FormPipeReader.cs
Comment (jkotalik) [reply src/Http/WebUtilities/src/FormPipeReader.cs]: You just don't need to call Advance. Save the position of the & and then find the =. Advance afterwards.

Comment (jkotalik) [reply src/Http/WebUtilities/src/FormPipeReader.cs]: Let's profile it though, my guess is saving the allocation for the sequence reader is worth it in the long run.

File: src/Http/WebUtilities/src/FormPipeReader.cs
Comment (Tratcher) [reply src/Http/WebUtilities/src/FormPipeReader.cs]: ? They're already shared class members.

File: src/Http/WebUtilities/src/FormPipeReader.cs
Comment (halter73) [reply src/Http/WebUtilities/src/FormPipeReader.cs]: Since it's a struct, you can just use assignment to create a copy of it on the stack and create a checkpoint of sorts. I think @jkotalik's suggestion of just not Advancing forward until you find the & and the = might be faster though.

Comment (Tratcher) [reply src/Http/WebUtilities/src/FormPipeReader.cs]: doesn't TryReadTo Advance by default? I'm only calling Advance in the negative case.

Comment (halter73) [reply src/Http/WebUtilities/src/FormPipeReader.cs]: Just looked at it. Sure seems like it does, and I don't see any easy way to tell it not to. Copying a SequenceReader at the beginning saves you a ctor call, but I doubt that has a huge perf impact relative to avoiding the extra struct altogether. I also don't think that helps you  limit your search for = to be before the next &.

Comment (jkotalik) [reply src/Http/WebUtilities/src/FormPipeReader.cs]: I think you are correct. There is a rewind method: https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System/Buffers/SequenceReader.cs#L156

File: src/Http/WebUtilities/src/FormPipeReader.cs
Comment (halter73) [inline src/Http/WebUtilities/src/FormPipeReader.cs]: Can we just call `segment` `sequence` instead?

File: src/Http/WebUtilities/src/FormPipeReader.cs
Comment (halter73) [reply src/Http/WebUtilities/src/FormPipeReader.cs]: Interesting. Might be worth benchmarking this vs the rewind solution. I could see the rewind solution being faster if the RetreatToPreviousSpan path is uncommon which I think it would be.

Comment (jkotalik) [reply src/Http/WebUtilities/src/FormPipeReader.cs]: Actually wouldn't we always retreat to the previous span as this case isn't within a single span?

File: src/Http/WebUtilities/src/FormPipeReader.cs
Comment (Tratcher) [reply src/Http/WebUtilities/src/FormPipeReader.cs]: Actually I do need to set it or it complains about using an uninitialized field later.

File: src/Http/WebUtilities/src/FormPipeReader.cs
Comment (Tratcher) [reply src/Http/WebUtilities/src/FormPipeReader.cs]: > Actually wouldn't we always retreat to the previous span as this case isn't within a single span?

Depends on where we find the "&"

File: src/Http/WebUtilities/src/FormPipeReader.cs
Comment (Tratcher) [reply src/Http/WebUtilities/src/FormPipeReader.cs]: I meant it as having isolated a block containing only one key-value pair... will rename.

**Comment by Tratcher** [comment]
Before:

Method |     Mean |     Error |    StdDev |      Op/s |  Gen 0 | Allocated |
-------------------------- |---------:|----------:|----------:|----------:|-------:|----------:|
ReadUtf8Data | 2.481 us | 0.0496 us | 0.0712 us | 403,029.6 | 0.0114 |     712 B |
ReadUtf8MultipleBlockData | 5.767 us | 0.1147 us | 0.1885 us | 173,402.3 | 0.0153 |     968 B | 

After:

Method |     Mean |     Error |    StdDev |      Op/s |  Gen 0 | Allocated |
-------------------------- |---------:|----------:|----------:|----------:|-------:|----------:|
ReadUtf8Data | 2.456 us | 0.0483 us | 0.0517 us | 407,244.3 | 0.0114 |     712 B |
ReadUtf8MultipleBlockData | 5.016 us | 0.0986 us | 0.2037 us | 199,379.4 | 0.0153 |     968 B |

LGTM

File: src/Http/WebUtilities/src/FormPipeReader.cs
Comment (halter73) [reply src/Http/WebUtilities/src/FormPipeReader.cs]: Maybe we shouldn't do rewind, it looks like it could make this algorithm O(n^2) in the pathological case since when you cross segments to rewind, it has to [rewalk the ReadOnlySequence](https://github.com/dotnet/corefx/blob/739f4437298d3d010f3427f557bf6d189aac67a7/src/System.Memory/src/System/Buffers/SequenceReader.cs#L173-L182) from the very beginning.

Comment (jkotalik) [reply src/Http/WebUtilities/src/FormPipeReader.cs]: Yeah, that isn't great. I wish there was a "TryPeekTo" method on SequenceReader s.t. we don't advance the reader. But agree with @halter73 here.

File: src/Http/WebUtilities/perf/Microsoft.AspNetCore.WebUtilities.Performance/FormPipeReaderInternalsBenchmark.cs
Comment (jkotalik) [inline src/Http/WebUtilities/perf/Microsoft.AspNetCore.WebUtilities.Performance/FormPipeReaderInternalsBenchmark.cs]: Maybe we should have a 3rd scenario which hits the fast path after the sequence reader finds a single segment span.

**Review by jkotalik — APPROVED** [review]
Good job!

**Review by halter73 — APPROVED** [review]
LGTM. I'd wait on @jkotalik's approval though

**Comment by jkotalik** [comment]
image

File: src/Http/WebUtilities/perf/Microsoft.AspNetCore.WebUtilities.Performance/FormPipeReaderInternalsBenchmark.cs
Comment (Tratcher) [reply src/Http/WebUtilities/perf/Microsoft.AspNetCore.WebUtilities.Performance/FormPipeReaderInternalsBenchmark.cs]: That's already covered. "foo=bar" will hit the fast path and then "baz=bo" will hit the slow path.

**Comment by anurse** [comment]
If we're taking this for Ask-Mode, needs the [template](https://github.com/aspnet/AspNetCore-Internal/wiki/.NET-Core-3-Ask-Mode) and label. Otherwise it needs to be on `master`.

**Comment by Tratcher** [comment]
@anurse added the template. Suggested for preview9 because we've had direct customer reports of the regression impacting their apps.

**Comment by anurse** [comment]
Sounds good. Meets the bar IMO, we'll confirm tomorrow.

**Comment by leecow** [comment]
Tactics approved

**Pull request merged**
